saleor/warehouse/availability.py

```diff
@@ -273,34 +290,145 @@ def check_preorder_threshold_bulk(
+    check_reservations: bool,
+):
+    """Validate if there is preorder available for given variants in given country.
+
+    It is used in orders, since it does not need additional logic related to limits.
+    :raises InsufficientStock: when there is not enough items in stock for a variant.
+    """
+    (
+        variants_channel_availability,
+        variants_global_allocations,
+        all_variants_channel_listings,
+        variant_channels,
+    ) = get_channel_data([variant], channel_slug)
 
     if check_reservations:
-        quantity_reservation_list = (
-            PreorderReservation.objects.filter(
-                product_variant_channel_listing__in=all_variants_channel_listings,
-                quantity_reserved__gt=0,
+        listings_reservations = _check_reservations(
+            checkout_lines, all_variants_channel_listings
+        )
+    else:
+        listings_reservations = defaultdict(int)
+
+    insufficient_stocks: List[InsufficientStockData] = []
+
+    if variants_channel_availability[variant.id][1] is not None:
+        channel_listing_id = variants_channel_availability[variant.id][2]
+        available_channel_quantity = variants_channel_availability[variant.id][0]
+        available_channel_quantity -= listings_reservations[channel_listing_id]
+        available_channel_quantity = max(available_channel_quantity, 0)
+
+        if quantity > available_channel_quantity:
+            insufficient_stocks.append(
+                InsufficientStockData(
+                    variant=variant, available_quantity=available_channel_quantity
+                )
             )
-            .not_expired()
-            .exclude_checkout_lines(checkout_lines)
-            .values("product_variant_channel_listing")
-            .annotate(quantity_reserved_sum=Sum("quantity_reserved"))
-        )  # type: ignore
-        listings_reservations: Dict = defaultdict(int)
-        for reservation in quantity_reservation_list:
-            listings_reservations[
-                reservation["product_variant_channel_listing"]
-            ] += reservation["quantity_reserved_sum"]
+
+        if variant.preorder_global_threshold is not None:
+            global_availability = variant.preorder_global_threshold
+            global_availability -= variants_global_allocations[variant.id]
+
+            for channel_listing in variant_channels[variant.id]:
+                global_availability -= listings_reservations[channel_listing.id]
+
+            global_availability = max(global_availability, 0)
+
+            if quantity > global_availability:
+                insufficient_stocks.append(
+                    InsufficientStockData(
+                        variant=variant,
+                        available_quantity=global_availability,
+                    )
+                )
+
+    if insufficient_stocks:
+        raise InsufficientStock(insufficient_stocks)
+
+
+def _check_reservations(
```

Comment:
Is this logic taken from the `reserve_preorders` function in `warehouse/reservations.py`? It looks identical - maybe it would be better to move this function to `reservations.py` module and reuse it in both places? Also the name should be more like "get_listings_reservations", it doesn't seem to check anything.

Reply:
TBH haven't even check `warehouse/reservations.py` - all i want to do is to just extract "reservation-ish" part from `check_preorder_bulk` so I can reuse it. But ofc I will look into `warehouse/reservations.py` :)

Reply:
You were right, now we have one `get_listings_reservations` used in `reserver_preorders` and in two functions in `availbility.py`